File: packages/case-definition-dsl/src/matchers/convenience/And.ts
import { COMBINE_MATCHERS_TYPE } from '@contract-case/case-entities-internal';
import { AnyMatcher } from '../base';
import { AnyMatcherOrData } from '../../types';

/**
 * Matches all of the provided matchers. Useful for combining restrictions
 * provided by different matchers, or creating new matchers without needing
 * plugins.
 *
 * For best results, wrap the And matcher in a WithExample matcher.
 *
 * @public
 */
export class And extends AnyMatcher {
  /** @internal */
  readonly '_case:matcher:type': typeof COMBINE_MATCHERS_TYPE;

  /** @internal */
  readonly '_case:matcher:children': Array<AnyMatcherOrData>;

  /**
   * @param matchers - An array of the matchers to run against this particular spot in the tree
   */
  constructor(matchers: AnyMatcherOrData[]) {
    super(COMBINE_MATCHERS_TYPE);
    this['_case:matcher:children'] = matchers;
  }

  /**
   * For non-TypeScript implementations (see `AnyMatcher.toJSON`)
   *
   * @privateRemarks
   * This comment and the implementation is boilerplate on all matchers to avoid
   * outputting duplicate unimportant documentation on all matcher classes of
   * the docs. Only modify this comment or the implementation via search and replace.
   */
  override toJSON(): unknown {
    return super.toJSON();
  }
}

File: packages/dsl-java/src/main/java/io/contract_testing/contractcase/dsl/matchers/convenience/And.java
package io.contract_testing.contractcase.dsl.matchers.convenience;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.contract_testing.contractcase.dsl.ContractCaseDsl;
import io.contract_testing.contractcase.dsl.DslMatcher;
import jakarta.annotation.Generated;
import java.lang.Object;
import java.lang.String;
import java.util.List;
import lombok.Builder;
import lombok.Getter;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

/**
 * Matches all of the provided matchers. Useful for combining restrictions
 * provided by different matchers, or creating new matchers without needing
 * plugins.<p>For best results, wrap the And matcher in a WithExample matcher.
 */
@Generated("@contract-case/case-definition-generator")
@ContractCaseDsl
public class And implements DslMatcher {

  /**
   * ContractCase's internal type for this element
   */
  @Getter
  @JsonProperty("_case:matcher:type")
  private final String type;

  /**
   * An array of the matchers to run against this particular spot in the tree
   */
  @Getter
  @JsonProperty("_case:matcher:children")
  private final List<Object> children;

  @Builder
  public And(@NotNull final List<Object> children) {
    this.type = "_case:And";
    this.children = children;
  }
}
